### PR TITLE
perf: enhancements

### PR DESCRIPTION
**Pull request by johncorrelli**

### CODE REVIEW

**Comment by johncorrelli** [comment]
Running this on TeamGantt staging environment, this change alone cuts the response time of the VERY LARGE requests in roughly half...

## Before
Screenshot_20201011_231513

## After
Screenshot_20201012_082841

File: spec/DayDifference.spec.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
+<?php
+
+use johncorrelli\dayDifference;
+use DateTime;
+
+describe('dayDifference', function() {
```
Comment (johncorrelli) [inline spec/DayDifference.spec.php]: `dayDifference` is for test to cover current name of the class. This gets correct in the next commit where the class is refactored and properly named.

File: src/DayDifference.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        $isFullWeek = count($allowedDaysOfTheWeek) === 7;
+        $hasExclusions = !empty($excludedDates);
+        $currentStamp = $startStamp;
+        $totalDays = 0;
+
+        while ($currentStamp < $endStamp) {
+            $totalDays += (int) (
+                ($isFullWeek || $this->isAllowed($currentStamp, $allowedDaysOfTheWeek))
+                && (!$hasExclusions || !$this->isExcluded($currentStamp, $excludedDates))
+            );
```
Comment (johncorrelli) [inline src/DayDifference.php]: this is quite a bit faster (in larger datasets) than

```php
if (is allowed and not excluded) {
    $totalDays++;
}
```

File: src/DayDifference.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    private function findDifference(int $startStamp, int $endStamp, array $allowedDaysOfTheWeek = [], array $excludedDates = []): int
+    {
+        $isFullWeek = count($allowedDaysOfTheWeek) === 7;
+        $hasExclusions = !empty($excludedDates);
+        $currentStamp = $startStamp;
+        $totalDays = 0;
+
+        while ($currentStamp < $endStamp) {
+            $totalDays += (int) (
+                ($isFullWeek || $this->isAllowed($currentStamp, $allowedDaysOfTheWeek))
```
Comment (bradbyte) [inline src/DayDifference.php:80]: Is this converting a bool to an int? Is that the part that is faster than the `var++`? 

What do you think about breaking this out into a variable like `$shouldIncrement` or something so the lines would read easier?

**Comment by bradbyte** [comment]
Very clean and well tested, just a question on one line.

File: src/DayDifference.php (hunk 3 above)
Comment (johncorrelli) [reply src/DayDifference.php:80]: Good question... I spent a lot of time on what these few lines are doing....

Some background:

```php
$a = (int) true; // = 1
$b = (int) false; // = 0
```

Essentially we're saying `$totalDays += 1` if the day is counted and `$totalDays += 0` if the day is excluded.

I initially had this as a check like you're suggesting, but it's surprising how much faster (on a large data set) not doing a variable assignment and an `if (should increment) { $totalDays++; }` is.

Perhaps I should add some comments in code to explain why?

Comment (bradbyte) [reply src/DayDifference.php:80]: Ah, that makes sense. I don't think you need comments, but if you renamed the variables or grouped them differently so it was more indicative of what that is doing, that could be helpful. 

I wasn't necessarily suggesting breaking it out beforehand, but just some thing like this...

```php
$isDayOfWeekAllowed =  ($isFullWeek || $this->isAllowed($currentStamp, $allowedDaysOfTheWeek));
$isDateAllowed = (!$hasExclusions || !$this->isExcluded($currentStamp, $excludedDates));

$incrementByOneOrZero = (int) ($isDayOfWeekAllowed && $isDateAllowed);

$totalDays += incrementByOneOrZero;
```

Comment (johncorrelli) [reply src/DayDifference.php:80]: This is wild... dropping that change into the test case on TG makes the response time jump pretty significantly....

Screenshot_20201012_101122

I did a multiple refreshes and the slowest was pretty much always between 12 and 14.

I think we're going to want to leave what we have... I added some comments to show whats up.

**Review by bradbyte — APPROVED** [review]

**Pull request merged**